{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Description</th>\n",
       "      <th>Gene_Accession_Number</th>\n",
       "      <th>data_1</th>\n",
       "      <th>call_1</th>\n",
       "      <th>data_2</th>\n",
       "      <th>call_2</th>\n",
       "      <th>data_3</th>\n",
       "      <th>call_3</th>\n",
       "      <th>data_4</th>\n",
       "      <th>call_4</th>\n",
       "      <th>...</th>\n",
       "      <th>data_29</th>\n",
       "      <th>call_29</th>\n",
       "      <th>data_30</th>\n",
       "      <th>call_30</th>\n",
       "      <th>data_31</th>\n",
       "      <th>call_31</th>\n",
       "      <th>data_32</th>\n",
       "      <th>call_32</th>\n",
       "      <th>data_33</th>\n",
       "      <th>call_33</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFFX-BioB-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-5_at</td>\n",
       "      <td>-214</td>\n",
       "      <td>A</td>\n",
       "      <td>-139</td>\n",
       "      <td>A</td>\n",
       "      <td>-76</td>\n",
       "      <td>A</td>\n",
       "      <td>-135</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>A</td>\n",
       "      <td>-318</td>\n",
       "      <td>A</td>\n",
       "      <td>-32</td>\n",
       "      <td>A</td>\n",
       "      <td>-124</td>\n",
       "      <td>A</td>\n",
       "      <td>-135</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFFX-BioB-M_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-M_at</td>\n",
       "      <td>-153</td>\n",
       "      <td>A</td>\n",
       "      <td>-73</td>\n",
       "      <td>A</td>\n",
       "      <td>-49</td>\n",
       "      <td>A</td>\n",
       "      <td>-114</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-114</td>\n",
       "      <td>A</td>\n",
       "      <td>-192</td>\n",
       "      <td>A</td>\n",
       "      <td>-49</td>\n",
       "      <td>A</td>\n",
       "      <td>-79</td>\n",
       "      <td>A</td>\n",
       "      <td>-186</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFFX-BioB-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-3_at</td>\n",
       "      <td>-58</td>\n",
       "      <td>A</td>\n",
       "      <td>-1</td>\n",
       "      <td>A</td>\n",
       "      <td>-307</td>\n",
       "      <td>A</td>\n",
       "      <td>265</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>-95</td>\n",
       "      <td>A</td>\n",
       "      <td>49</td>\n",
       "      <td>A</td>\n",
       "      <td>-37</td>\n",
       "      <td>A</td>\n",
       "      <td>-70</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AFFX-BioC-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-5_at</td>\n",
       "      <td>88</td>\n",
       "      <td>A</td>\n",
       "      <td>283</td>\n",
       "      <td>A</td>\n",
       "      <td>309</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>193</td>\n",
       "      <td>A</td>\n",
       "      <td>312</td>\n",
       "      <td>A</td>\n",
       "      <td>230</td>\n",
       "      <td>P</td>\n",
       "      <td>330</td>\n",
       "      <td>A</td>\n",
       "      <td>337</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFFX-BioC-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-3_at</td>\n",
       "      <td>-295</td>\n",
       "      <td>A</td>\n",
       "      <td>-264</td>\n",
       "      <td>A</td>\n",
       "      <td>-376</td>\n",
       "      <td>A</td>\n",
       "      <td>-419</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-51</td>\n",
       "      <td>A</td>\n",
       "      <td>-139</td>\n",
       "      <td>A</td>\n",
       "      <td>-367</td>\n",
       "      <td>A</td>\n",
       "      <td>-188</td>\n",
       "      <td>A</td>\n",
       "      <td>-407</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene_Description Gene_Accession_Number  data_1 call_1  \\\n",
       "0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at    -214      A   \n",
       "1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at    -153      A   \n",
       "2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at     -58      A   \n",
       "3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at      88      A   \n",
       "4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at    -295      A   \n",
       "\n",
       "   data_2 call_2  data_3 call_3  data_4 call_4   ...    data_29 call_29  \\\n",
       "0    -139      A     -76      A    -135      A   ...         15       A   \n",
       "1     -73      A     -49      A    -114      A   ...       -114       A   \n",
       "2      -1      A    -307      A     265      A   ...          2       A   \n",
       "3     283      A     309      A      12      A   ...        193       A   \n",
       "4    -264      A    -376      A    -419      A   ...        -51       A   \n",
       "\n",
       "   data_30 call_30  data_31 call_31  data_32 call_32  data_33 call_33  \n",
       "0     -318       A      -32       A     -124       A     -135       A  \n",
       "1     -192       A      -49       A      -79       A     -186       A  \n",
       "2      -95       A       49       A      -37       A      -70       A  \n",
       "3      312       A      230       P      330       A      337       A  \n",
       "4     -139       A     -367       A     -188       A     -407       A  \n",
       "\n",
       "[5 rows x 78 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Goal: implement differential expression analysis\n",
    "for two Leukemia conditions usig the Z-test.\n",
    "\n",
    "Notes:\n",
    "\n",
    "- The data for this analysis was downloaded from\n",
    "    http://portals.broadinstitute.org/cgi-bin/cancer/publications/pub_paper.cgi?mode=view&paper_id=43\n",
    "- This code is an adaptation for pandas of the code provided at\n",
    "    http://dept.stat.lsa.umich.edu/~kshedden/Python-Workshop/gene_expression_comparison.html\n",
    "    which uses data in \"SOFT\" format -- https://www.ncbi.nlm.nih.gov/sites/GDSbrowser?acc=GDS1615\n",
    "\n",
    "\n",
    "@author: Andrei Sura\n",
    "@see https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3134237/ - Gene Set Enrichment Analysis Made Simple\n",
    "@see http://www.gettinggeneticsdone.com/2012/03/pathway-analysis-for-high-throughput.html\n",
    "\n",
    "#########################################################\n",
    "# Expression profiling in early onset colorectal cancer\n",
    "@see http://clincancerres.aacrjournals.org/content/13/4/1107.full-text.pdf\n",
    "@see https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE4107\n",
    "\n",
    "@see http://isites.harvard.edu/fs/docs/icb.topic1517419.files/golub_analysis.R\n",
    "@see http://svitsrv25.epfl.ch/R-doc/library/multtest/doc/golub.R\n",
    "\n",
    "@see https://www.biostars.org/p/16137/\n",
    "@see http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0012336\n",
    "@see http://bioinformatics.mdanderson.org/MicroarrayCourse/Lectures09/ma10b.pdf\n",
    "@see http://sebastianraschka.com/Articles/2014_pca_step_by_step.html\n",
    "\n",
    "@see https://www.biostars.org/p/42219/ - pathway analysis\n",
    "\"\"\"\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats.distributions import norm\n",
    "# np.seterr(invalid='ignore')\n",
    "\n",
    "IN_FILE = '/Users/asura/git/gene_expression_two_group_comparison/_golub_assignment/data_set_ALL_AML_train.csv'\n",
    "OUT_FILE = '/Users/asura/git/gene_expression_two_group_comparison/_golub_assignment/out.csv'\n",
    "\n",
    "# RANGE_ALL = [\"call.{}\".format(x) for x in range(0, 27)]\n",
    "# RANGE_AML = [\"call.{}\".format(x) for x in range(27, 38)]\n",
    "RANGE_ALL = [\"data_{}\".format(x) for x in range(1, 28)]\n",
    "RANGE_AML = [\"data_{}\".format(x) for x in range(28, 39)]\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "38 samples divided in two groups:\n",
    "- ALL: 27 (1 to 27)  \n",
    "- AML: 11 (28 to 38)\n",
    "\n",
    "An accession number in bioinformatics is a unique identifier given to a DNA\n",
    "or protein sequence record to allow for tracking of different versions of\n",
    "that sequence record and the associated sequence over time in a single data\n",
    "repository. Because of its relative stability, accession numbers can be \n",
    "utilized as foreign keys for referring to a sequence object, but not necessarily\n",
    "to a unique sequence. All sequence information repositories implement the\n",
    "concept of \"accession number\" but might do so with subtle variations.\n",
    "\n",
    "\"\"\"\n",
    "df = pd.read_csv(IN_FILE, sep='\\t', index_col=False, skipinitialspace=False, nrows=None)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Description</th>\n",
       "      <th>data_1</th>\n",
       "      <th>data_2</th>\n",
       "      <th>data_3</th>\n",
       "      <th>data_4</th>\n",
       "      <th>data_5</th>\n",
       "      <th>data_6</th>\n",
       "      <th>data_7</th>\n",
       "      <th>data_8</th>\n",
       "      <th>data_9</th>\n",
       "      <th>...</th>\n",
       "      <th>data_29</th>\n",
       "      <th>data_30</th>\n",
       "      <th>data_31</th>\n",
       "      <th>data_32</th>\n",
       "      <th>data_33</th>\n",
       "      <th>data_34</th>\n",
       "      <th>data_35</th>\n",
       "      <th>data_36</th>\n",
       "      <th>data_37</th>\n",
       "      <th>data_38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFFX-BioB-5_at (endogenous control)</td>\n",
       "      <td>-214</td>\n",
       "      <td>-139</td>\n",
       "      <td>-76</td>\n",
       "      <td>-135</td>\n",
       "      <td>-106</td>\n",
       "      <td>-138</td>\n",
       "      <td>-72</td>\n",
       "      <td>-413</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>-318</td>\n",
       "      <td>-32</td>\n",
       "      <td>-124</td>\n",
       "      <td>-135</td>\n",
       "      <td>-20</td>\n",
       "      <td>7</td>\n",
       "      <td>-213</td>\n",
       "      <td>-25</td>\n",
       "      <td>-72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFFX-BioB-M_at (endogenous control)</td>\n",
       "      <td>-153</td>\n",
       "      <td>-73</td>\n",
       "      <td>-49</td>\n",
       "      <td>-114</td>\n",
       "      <td>-125</td>\n",
       "      <td>-85</td>\n",
       "      <td>-144</td>\n",
       "      <td>-260</td>\n",
       "      <td>-127</td>\n",
       "      <td>...</td>\n",
       "      <td>-114</td>\n",
       "      <td>-192</td>\n",
       "      <td>-49</td>\n",
       "      <td>-79</td>\n",
       "      <td>-186</td>\n",
       "      <td>-207</td>\n",
       "      <td>-100</td>\n",
       "      <td>-252</td>\n",
       "      <td>-20</td>\n",
       "      <td>-139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFFX-BioB-3_at (endogenous control)</td>\n",
       "      <td>-58</td>\n",
       "      <td>-1</td>\n",
       "      <td>-307</td>\n",
       "      <td>265</td>\n",
       "      <td>-76</td>\n",
       "      <td>215</td>\n",
       "      <td>238</td>\n",
       "      <td>7</td>\n",
       "      <td>106</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>-95</td>\n",
       "      <td>49</td>\n",
       "      <td>-37</td>\n",
       "      <td>-70</td>\n",
       "      <td>-50</td>\n",
       "      <td>-57</td>\n",
       "      <td>136</td>\n",
       "      <td>124</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AFFX-BioC-5_at (endogenous control)</td>\n",
       "      <td>88</td>\n",
       "      <td>283</td>\n",
       "      <td>309</td>\n",
       "      <td>12</td>\n",
       "      <td>168</td>\n",
       "      <td>71</td>\n",
       "      <td>55</td>\n",
       "      <td>-2</td>\n",
       "      <td>268</td>\n",
       "      <td>...</td>\n",
       "      <td>193</td>\n",
       "      <td>312</td>\n",
       "      <td>230</td>\n",
       "      <td>330</td>\n",
       "      <td>337</td>\n",
       "      <td>101</td>\n",
       "      <td>132</td>\n",
       "      <td>318</td>\n",
       "      <td>325</td>\n",
       "      <td>392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFFX-BioC-3_at (endogenous control)</td>\n",
       "      <td>-295</td>\n",
       "      <td>-264</td>\n",
       "      <td>-376</td>\n",
       "      <td>-419</td>\n",
       "      <td>-230</td>\n",
       "      <td>-272</td>\n",
       "      <td>-399</td>\n",
       "      <td>-541</td>\n",
       "      <td>-210</td>\n",
       "      <td>...</td>\n",
       "      <td>-51</td>\n",
       "      <td>-139</td>\n",
       "      <td>-367</td>\n",
       "      <td>-188</td>\n",
       "      <td>-407</td>\n",
       "      <td>-369</td>\n",
       "      <td>-377</td>\n",
       "      <td>-209</td>\n",
       "      <td>-396</td>\n",
       "      <td>-324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene_Description  data_1  data_2  data_3  data_4  \\\n",
       "0  AFFX-BioB-5_at (endogenous control)    -214    -139     -76    -135   \n",
       "1  AFFX-BioB-M_at (endogenous control)    -153     -73     -49    -114   \n",
       "2  AFFX-BioB-3_at (endogenous control)     -58      -1    -307     265   \n",
       "3  AFFX-BioC-5_at (endogenous control)      88     283     309      12   \n",
       "4  AFFX-BioC-3_at (endogenous control)    -295    -264    -376    -419   \n",
       "\n",
       "   data_5  data_6  data_7  data_8  data_9   ...     data_29  data_30  data_31  \\\n",
       "0    -106    -138     -72    -413       5   ...          15     -318      -32   \n",
       "1    -125     -85    -144    -260    -127   ...        -114     -192      -49   \n",
       "2     -76     215     238       7     106   ...           2      -95       49   \n",
       "3     168      71      55      -2     268   ...         193      312      230   \n",
       "4    -230    -272    -399    -541    -210   ...         -51     -139     -367   \n",
       "\n",
       "   data_32  data_33  data_34  data_35  data_36  data_37  data_38  \n",
       "0     -124     -135      -20        7     -213      -25      -72  \n",
       "1      -79     -186     -207     -100     -252      -20     -139  \n",
       "2      -37      -70      -50      -57      136      124       -1  \n",
       "3      330      337      101      132      318      325      392  \n",
       "4     -188     -407     -369     -377     -209     -396     -324  \n",
       "\n",
       "[5 rows x 39 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = [\"Gene_Description\"]\n",
    "columns.extend(RANGE_ALL)\n",
    "columns.extend(RANGE_AML)\n",
    "\n",
    "columns_data = []\n",
    "columns_data.extend(RANGE_ALL)\n",
    "columns_data.extend(RANGE_AML)\n",
    "\n",
    "df = df[columns]\n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>data_28</th>\n",
       "      <th>data_29</th>\n",
       "      <th>data_30</th>\n",
       "      <th>data_31</th>\n",
       "      <th>data_32</th>\n",
       "      <th>data_33</th>\n",
       "      <th>data_34</th>\n",
       "      <th>data_35</th>\n",
       "      <th>data_36</th>\n",
       "      <th>data_37</th>\n",
       "      <th>data_38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-4</td>\n",
       "      <td>15</td>\n",
       "      <td>-318</td>\n",
       "      <td>-32</td>\n",
       "      <td>-124</td>\n",
       "      <td>-135</td>\n",
       "      <td>-20</td>\n",
       "      <td>7</td>\n",
       "      <td>-213</td>\n",
       "      <td>-25</td>\n",
       "      <td>-72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-116</td>\n",
       "      <td>-114</td>\n",
       "      <td>-192</td>\n",
       "      <td>-49</td>\n",
       "      <td>-79</td>\n",
       "      <td>-186</td>\n",
       "      <td>-207</td>\n",
       "      <td>-100</td>\n",
       "      <td>-252</td>\n",
       "      <td>-20</td>\n",
       "      <td>-139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-125</td>\n",
       "      <td>2</td>\n",
       "      <td>-95</td>\n",
       "      <td>49</td>\n",
       "      <td>-37</td>\n",
       "      <td>-70</td>\n",
       "      <td>-50</td>\n",
       "      <td>-57</td>\n",
       "      <td>136</td>\n",
       "      <td>124</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>241</td>\n",
       "      <td>193</td>\n",
       "      <td>312</td>\n",
       "      <td>230</td>\n",
       "      <td>330</td>\n",
       "      <td>337</td>\n",
       "      <td>101</td>\n",
       "      <td>132</td>\n",
       "      <td>318</td>\n",
       "      <td>325</td>\n",
       "      <td>392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-191</td>\n",
       "      <td>-51</td>\n",
       "      <td>-139</td>\n",
       "      <td>-367</td>\n",
       "      <td>-188</td>\n",
       "      <td>-407</td>\n",
       "      <td>-369</td>\n",
       "      <td>-377</td>\n",
       "      <td>-209</td>\n",
       "      <td>-396</td>\n",
       "      <td>-324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   data_28  data_29  data_30  data_31  data_32  data_33  data_34  data_35  \\\n",
       "0       -4       15     -318      -32     -124     -135      -20        7   \n",
       "1     -116     -114     -192      -49      -79     -186     -207     -100   \n",
       "2     -125        2      -95       49      -37      -70      -50      -57   \n",
       "3      241      193      312      230      330      337      101      132   \n",
       "4     -191      -51     -139     -367     -188     -407     -369     -377   \n",
       "\n",
       "   data_36  data_37  data_38  \n",
       "0     -213      -25      -72  \n",
       "1     -252      -20     -139  \n",
       "2      136      124       -1  \n",
       "3      318      325      392  \n",
       "4     -209     -396     -324  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all = df[RANGE_ALL]\n",
    "df_aml = df[RANGE_AML]\n",
    "df_aml.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Z mean: 0.23653755413102437\n",
      "Z std: 1.7866583897611175\n",
      "Z means: 0.25866180389956517, 0.15948940945434142, 0.8405105905456586\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# df_all = df_all.applymap(lambda x: np.log(x) / np.log(2))\n",
    "# df_aml = df_aml.applymap(lambda x: np.log(x) / np.log(2))\n",
    "\n",
    "mean_all = df_all.mean(axis=1)  # means for 7129 genes\n",
    "mean_aml = df_aml.mean(axis=1)   \n",
    "var_all = df_all.var(axis=1)  # variance for 7129 genes\n",
    "var_aml = df_aml.var(axis=1)\n",
    "\n",
    "num_all = len(RANGE_ALL)  # number of samples of each condition\n",
    "num_aml = len(RANGE_AML)\n",
    "\n",
    "Z = (mean_all - mean_aml) / np.sqrt(var_all/num_all + var_aml/num_aml)\n",
    "# print(\"means ALL: {}\\n {}\".format(len(mean_all), mean_all))\n",
    "# print(\"means AML: {}\\n {}\".format(len(mean_aml), mean_aml))\n",
    "\n",
    "print(\"Z mean: {}\".format(Z.mean()))\n",
    "print(\"Z std: {}\".format(Z.std()))\n",
    "\n",
    "\"\"\"\n",
    "Z mean: 0.14004765578280895\n",
    "Z std: 1.6058829875390022\n",
    "Z means: 0.22008696871931546, 0.13073362322906437, 0.8692663767709357\n",
    "\n",
    "Since the standard deviation is greater than 1, there appear to be multiple\n",
    "genes for which the mean expression levels in the ALL and AML samples differ. \n",
    "Further, since the mean Z-score is positive, it appears that the dominant pattern\n",
    "is for genes to be expressed at a higher level in the ALL compared to the AML samples.\n",
    "\"\"\"\n",
    "\n",
    "mean_z_mod = np.mean(np.abs(Z) > 2)\n",
    "mean_z1 = np.mean(Z > 2)\n",
    "mean_z2 = np.mean(Z < 2)\n",
    "\n",
    "print(\"Z means: {}, {}, {}\".format(mean_z_mod, mean_z1, mean_z2))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Check the mean of values with low Z magnitude: 0.2631578947368421\n",
      "Check the mean of values with high Z magnitude: 0.10526315789473684\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Genes with low expression level are harder to measure accurately, \n",
    "thus we expect that fewer of these genes will meet a given statistical\n",
    "threshold for differential expression. Similarly, genes with low variance\n",
    "are potentially less likely to be associated with biological differences.\n",
    "We can assess these trends in our data. First, we will determine which\n",
    "genes are in the lower and upper halves of all genes based on either \n",
    "mean expression or expression variation.\n",
    "\"\"\"\n",
    "SD = df[columns_data].std()\n",
    "index_stdev_low = np.flatnonzero(SD < np.median(SD))\n",
    "index_stdev_high = np.flatnonzero(SD > np.median(SD))\n",
    "\n",
    "MEAN = df[columns_data].mean()\n",
    "index_mean_low = np.flatnonzero(MEAN < np.median(MEAN))\n",
    "index_mean_high = np.flatnonzero(MEAN > np.median(MEAN))\n",
    "\n",
    "\"\"\"\n",
    "Now we can look at the proportion of genes within each of these\n",
    "strata that have Z-score magnitude greater than two.\n",
    "\"\"\"\n",
    "mean_low = np.mean(np.abs(Z[index_stdev_low]) > 2)\n",
    "mean_high = np.mean(np.abs(Z[index_stdev_high]) > 2)\n",
    "\n",
    "print(\"Check the mean of values with low Z magnitude: {}\".format(mean_low))\n",
    "print(\"Check the mean of values with high Z magnitude: {}\".format(mean_high))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Z-score threshold under a Bonferroni correction: 4.4932751673227616\n",
      "Found 111 genes with Z > 4.4932751673227616\n",
      "\n",
      "148 - Clone 22 mRNA, alternative splice variant alpha-1\n",
      "228 - KIAA0102 gene\n",
      "378 - AARS Alanyl-tRNA synthetase\n",
      "435 - KIAA0097 gene\n",
      "437 - T-COMPLEX PROTEIN 1, EPSILON SUBUNIT\n",
      "460 - Liver mRNA for interferon-gamma inducing factor(IGIF)\n",
      "531 - HMG1 High-mobility group (nonhistone chromosomal) protein 1\n",
      "537 - KIAA0159 gene\n",
      "653 - Non-lens beta gamma-crystallin like protein (AIM1) mRNA, partial cds\n",
      "667 - KIAA0212 gene\n",
      "696 - KIAA0235 gene, partial cds\n",
      "803 - Macmarcks\n",
      "877 - Heat Shock Protein, 70 Kda (Gb:Y00371)\n",
      "1077 - ADPRT ADP-ribosyltransferase (NAD+; poly (ADP-ribose) polymerase)\n",
      "1080 - UBIQUITIN-LIKE PROTEIN GDX\n",
      "1143 - SPTAN1 Spectrin, alpha, non-erythrocytic 1 (alpha-fodrin)\n",
      "1206 - Protein tyrosine kinase related mRNA sequence\n",
      "1240 - IEF SSP 9502 mRNA\n",
      "1244 - GBE1 Glucan (1,4-alpha-), branching enzyme 1 (glycogen branching enzyme, Andersen disease, glycogen storage disease type IV)\n",
      "1248 - INDUCED MYELOID LEUKEMIA CELL DIFFERENTIATION PROTEIN MCL1\n",
      "1305 - CRYZ Crystallin zeta (quinone reductase)\n",
      "1620 - Motor protein\n",
      "1629 - Inducible protein mRNA\n",
      "1673 - FTL Ferritin, light polypeptide\n",
      "1684 - Terminal transferase mRNA\n",
      "1703 - ADA Adenosine deaminase\n",
      "1744 - LYN V-yes-1 Yamaguchi sarcoma viral related oncogene homolog\n",
      "1806 - Neuromedin B mRNA\n",
      "1826 - TP53 Tumor protein p53 (Li-Fraumeni syndrome)\n",
      "1833 - CD33 CD33 antigen (differentiation antigen)\n",
      "1897 - LIPC Lipase, hepatic\n",
      "1903 - CARCINOEMBRYONIC ANTIGEN PRECURSOR\n",
      "1908 - IL7R Interleukin 7 receptor\n",
      "1927 - Oncoprotein 18 (Op18) gene\n",
      "2000 - CSF1 Colony-stimulating factor 1 (M-CSF)\n",
      "2019 - FAH Fumarylacetoacetate\n",
      "2049 - ZNF42 Zinc finger protein 42 (myeloid-specific retinoic acid-responsive)\n",
      "2083 - DCK Deoxycytidine kinase\n",
      "2110 - ATP6C Vacuolar H+ ATPase proton channel subunit\n",
      "2130 - RPA1 Replication protein A1 (70kD)\n",
      "2232 - NUCLEOLYSIN TIA-1\n",
      "2241 - PEPTIDYL-PROLYL CIS-TRANS ISOMERASE, MITOCHONDRIAL PRECURSOR\n",
      "2266 - CDC25A Cell division cycle 25A\n",
      "2300 - ACTN2 Actinin alpha 2 \n",
      "2334 - IGB Immunoglobulin-associated beta (B29)\n",
      "2347 - ACADM Acyl-Coenzyme A dehydrogenase, C-4 to C-12 straight chain\n",
      "2353 - CCND3 Cyclin D3\n",
      "2440 - HKR-T1\n",
      "2496 - BLK Protein-tyrosine kinase blk\n",
      "2625 - MSH2 DNA repair protein MSH2\n",
      "2641 - MB-1 gene\n",
      "2758 - Thrombospondin-p50 gene extracted from Human thrombospondin-1 gene, partial cds\n",
      "2908 - SRP9 Signal recognition particle 9 kD protein\n",
      "3055 - Cytoplasmic dynein light chain 1 (hdlc1) mRNA\n",
      "3069 - Tax1-binding protein TXBP181 mRNA\n",
      "3095 - Heterochromatin protein p25 mRNA\n",
      "3136 - Stimulator of TAR RNA binding (SRB) mRNA\n",
      "3311 - Protein kinase ATR mRNA\n",
      "3319 - Leukotriene C4 synthase (LTC4S) gene\n",
      "3390 - NADH:ubiquinone oxidoreductase subunit B13 (B13) mRNA\n",
      "3506 - Putative enterocyte differentiation promoting factor mRNA, partial cds\n",
      "3560 - JUN V-jun avian sarcoma virus 17 oncogene homolog\n",
      "3659 - PLATELET-ACTIVATING FACTOR ACETYLHYDROLASE 45 KD SUBUNIT\n",
      "3846 - GB DEF = Homeodomain protein HoxA9 mRNA\n",
      "3898 - Serine kinase SRPK2 mRNA\n",
      "4051 - Catalase (EC 1.11.1.6) 5'flank and exon 1 mapping to chromosome 11, band p13 (and joined CDS)\n",
      "4057 - GJB1 Gap junction protein, beta 1, 32kD (connexin 32, Charcot-Marie-Tooth neuropathy, X-linked)\n",
      "4152 - HISTONE H2A.X\n",
      "4166 - ALDR1 Aldehyde reductase 1 (low Km aldose reductase)\n",
      "4176 - IRF2 Interferon regulatory factor 2\n",
      "4195 - PRG1 Proteoglycan 1, secretory granule\n",
      "4210 - VIL2 Villin 2 (ezrin)\n",
      "4229 - CTPS CTP synthetase\n",
      "4288 - ADH4 gene for class II alcohol dehydrogenase (pi subunit), exon 1\n",
      "4327 - PROTEASOME IOTA CHAIN\n",
      "4367 - PROTEASOME ZETA CHAIN\n",
      "4374 - DAGK1 Diacylglycerol kinase, alpha (80kD)\n",
      "4388 - GTF2E2 General transcription factor TFIIE beta subunit, 34 kD\n",
      "4437 - LMP2 gene extracted from H.sapiens genes TAP1, TAP2, LMP2, LMP7 and DOB\n",
      "4447 - EWSR1 Ewing sarcoma breakpoint region 1\n",
      "4534 - RETINOBLASTOMA BINDING PROTEIN P48\n",
      "4568 - P130 mRNA for 130K protein\n",
      "4581 - GLRX Glutaredoxin (thioltransferase)\n",
      "4846 - Zyxin\n",
      "4935 - GPX1 Glutathione peroxidase 1\n",
      "4950 - Nucleoside-diphosphate kinase\n",
      "4972 - RABAPTIN-5 protein\n",
      "5038 - LEPR Leptin receptor\n",
      "5071 - Cytokeratin 17\n",
      "5074 - EEF1D Eukaryotic translation elongation factor 1 delta (guanine nucleotide exchange protein)\n",
      "5190 - Adenosine triphosphatase, calcium\n",
      "5253 - MCM3 Minichromosome maintenance deficient (S. cerevisiae) 3\n",
      "5347 - CYP2C18 Cytochrome P450, subfamily IIC (mephenytoin 4-hydroxylase), polypeptide 18\n",
      "5500 - TOP2B Topoisomerase (DNA) II beta (180kD)\n",
      "5529 - Integrin cytoplasmic domain associated protein (Icap-1a) mRNA\n",
      "5551 - PROBABLE G PROTEIN-COUPLED RECEPTOR LCR1 HOMOLOG\n",
      "5592 - Transcriptional activator hSNF2b\n",
      "5771 - C-myb gene extracted from Human (c-myb) gene, complete primary cds, and five complete alternatively spliced cds\n",
      "6224 - CD19 gene\n",
      "6280 - MYL1 Myosin light chain (alkali)\n",
      "6282 - TCF3 Transcription factor 3 (E2A immunoglobulin enhancer binding factors E12/E47)\n",
      "6470 - Thymopoietin beta mRNA\n",
      "6514 - DHPS Deoxyhypusine synthase\n",
      "6643 - Butyrophilin (BTF5) mRNA\n",
      "6676 - HOX 2.2 gene extracted from Human Hox2.2 gene for a homeobox protein\n",
      "6701 - LPAP gene\n",
      "6800 - GB DEF = Retinoblastoma susceptibility protein (RB1) gene, with a 3 bp deletion in exon 22 (L11910 bases 161855-162161)\n",
      "6854 - TCF3 Transcription factor 3 (E2A immunoglobulin enhancer binding factors E12/E47)\n",
      "6894 - Uridine diphosphoglucose pyrophosphorylase mRNA\n",
      "6973 - CD19 CD19 antigen\n",
      "7118 - Transcriptional activator hSNF2b\n"
     ]
    }
   ],
   "source": [
    "## The Z-score threshold under a Bonferroni correction\n",
    "zst = -norm.ppf(0.025/Z.shape[0])\n",
    "indexes = np.flatnonzero(np.abs(Z) > zst)\n",
    "\n",
    "print(\"Z-score threshold under a Bonferroni correction: {}\".format(zst))\n",
    "print(\"Found {} genes with Z > {}\\n\".format(len(indexes), zst))\n",
    "\n",
    "with open(OUT_FILE, \"w\") as fout:\n",
    "    for i in indexes:\n",
    "        # print(\"Found gene: {} - {}\".format(i, df[\"Gene Description\"][i]))\n",
    "        print(\"{} - {}\".format(i, df[\"Gene_Description\"][i]))\n",
    "        fout.write(\"{},{}\\n\".format(Z[i], df[\"Gene_Description\"][i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
