{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Description</th>\n",
       "      <th>call_33</th>\n",
       "      <th>data_1</th>\n",
       "      <th>call_1</th>\n",
       "      <th>data_2</th>\n",
       "      <th>call_2</th>\n",
       "      <th>data_3</th>\n",
       "      <th>call_3</th>\n",
       "      <th>data_4</th>\n",
       "      <th>call_4</th>\n",
       "      <th>...</th>\n",
       "      <th>data_29</th>\n",
       "      <th>call_29</th>\n",
       "      <th>data_30</th>\n",
       "      <th>call_30</th>\n",
       "      <th>data_31</th>\n",
       "      <th>call_31</th>\n",
       "      <th>data_32</th>\n",
       "      <th>call_32</th>\n",
       "      <th>data_33</th>\n",
       "      <th>descr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AFFX-BioB-5_at (endogenous control)</th>\n",
       "      <td>AFFX-BioB-5_at</td>\n",
       "      <td>-214</td>\n",
       "      <td>A</td>\n",
       "      <td>-139</td>\n",
       "      <td>A</td>\n",
       "      <td>-76</td>\n",
       "      <td>A</td>\n",
       "      <td>-135</td>\n",
       "      <td>A</td>\n",
       "      <td>-106</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>-318</td>\n",
       "      <td>A</td>\n",
       "      <td>-32</td>\n",
       "      <td>A</td>\n",
       "      <td>-124</td>\n",
       "      <td>A</td>\n",
       "      <td>-135</td>\n",
       "      <td>A</td>\n",
       "      <td>AFFX-BioB-5_at (endogenous control)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioB-M_at (endogenous control)</th>\n",
       "      <td>AFFX-BioB-M_at</td>\n",
       "      <td>-153</td>\n",
       "      <td>A</td>\n",
       "      <td>-73</td>\n",
       "      <td>A</td>\n",
       "      <td>-49</td>\n",
       "      <td>A</td>\n",
       "      <td>-114</td>\n",
       "      <td>A</td>\n",
       "      <td>-125</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>-192</td>\n",
       "      <td>A</td>\n",
       "      <td>-49</td>\n",
       "      <td>A</td>\n",
       "      <td>-79</td>\n",
       "      <td>A</td>\n",
       "      <td>-186</td>\n",
       "      <td>A</td>\n",
       "      <td>AFFX-BioB-M_at (endogenous control)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioB-3_at (endogenous control)</th>\n",
       "      <td>AFFX-BioB-3_at</td>\n",
       "      <td>-58</td>\n",
       "      <td>A</td>\n",
       "      <td>-1</td>\n",
       "      <td>A</td>\n",
       "      <td>-307</td>\n",
       "      <td>A</td>\n",
       "      <td>265</td>\n",
       "      <td>A</td>\n",
       "      <td>-76</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>-95</td>\n",
       "      <td>A</td>\n",
       "      <td>49</td>\n",
       "      <td>A</td>\n",
       "      <td>-37</td>\n",
       "      <td>A</td>\n",
       "      <td>-70</td>\n",
       "      <td>A</td>\n",
       "      <td>AFFX-BioB-3_at (endogenous control)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioC-5_at (endogenous control)</th>\n",
       "      <td>AFFX-BioC-5_at</td>\n",
       "      <td>88</td>\n",
       "      <td>A</td>\n",
       "      <td>283</td>\n",
       "      <td>A</td>\n",
       "      <td>309</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>A</td>\n",
       "      <td>168</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>312</td>\n",
       "      <td>A</td>\n",
       "      <td>230</td>\n",
       "      <td>P</td>\n",
       "      <td>330</td>\n",
       "      <td>A</td>\n",
       "      <td>337</td>\n",
       "      <td>A</td>\n",
       "      <td>AFFX-BioC-5_at (endogenous control)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioC-3_at (endogenous control)</th>\n",
       "      <td>AFFX-BioC-3_at</td>\n",
       "      <td>-295</td>\n",
       "      <td>A</td>\n",
       "      <td>-264</td>\n",
       "      <td>A</td>\n",
       "      <td>-376</td>\n",
       "      <td>A</td>\n",
       "      <td>-419</td>\n",
       "      <td>A</td>\n",
       "      <td>-230</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>-139</td>\n",
       "      <td>A</td>\n",
       "      <td>-367</td>\n",
       "      <td>A</td>\n",
       "      <td>-188</td>\n",
       "      <td>A</td>\n",
       "      <td>-407</td>\n",
       "      <td>A</td>\n",
       "      <td>AFFX-BioC-3_at (endogenous control)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Gene_Description  call_33 data_1  call_1  \\\n",
       "AFFX-BioB-5_at (endogenous control)   AFFX-BioB-5_at     -214      A    -139   \n",
       "AFFX-BioB-M_at (endogenous control)   AFFX-BioB-M_at     -153      A     -73   \n",
       "AFFX-BioB-3_at (endogenous control)   AFFX-BioB-3_at      -58      A      -1   \n",
       "AFFX-BioC-5_at (endogenous control)   AFFX-BioC-5_at       88      A     283   \n",
       "AFFX-BioC-3_at (endogenous control)   AFFX-BioC-3_at     -295      A    -264   \n",
       "\n",
       "                                    data_2  call_2 data_3  call_3 data_4  \\\n",
       "AFFX-BioB-5_at (endogenous control)      A     -76      A    -135      A   \n",
       "AFFX-BioB-M_at (endogenous control)      A     -49      A    -114      A   \n",
       "AFFX-BioB-3_at (endogenous control)      A    -307      A     265      A   \n",
       "AFFX-BioC-5_at (endogenous control)      A     309      A      12      A   \n",
       "AFFX-BioC-3_at (endogenous control)      A    -376      A    -419      A   \n",
       "\n",
       "                                     call_4  \\\n",
       "AFFX-BioB-5_at (endogenous control)    -106   \n",
       "AFFX-BioB-M_at (endogenous control)    -125   \n",
       "AFFX-BioB-3_at (endogenous control)     -76   \n",
       "AFFX-BioC-5_at (endogenous control)     168   \n",
       "AFFX-BioC-3_at (endogenous control)    -230   \n",
       "\n",
       "                                                    ...                   \\\n",
       "AFFX-BioB-5_at (endogenous control)                 ...                    \n",
       "AFFX-BioB-M_at (endogenous control)                 ...                    \n",
       "AFFX-BioB-3_at (endogenous control)                 ...                    \n",
       "AFFX-BioC-5_at (endogenous control)                 ...                    \n",
       "AFFX-BioC-3_at (endogenous control)                 ...                    \n",
       "\n",
       "                                    data_29  call_29 data_30  call_30 data_31  \\\n",
       "AFFX-BioB-5_at (endogenous control)       A     -318       A      -32       A   \n",
       "AFFX-BioB-M_at (endogenous control)       A     -192       A      -49       A   \n",
       "AFFX-BioB-3_at (endogenous control)       A      -95       A       49       A   \n",
       "AFFX-BioC-5_at (endogenous control)       A      312       A      230       P   \n",
       "AFFX-BioC-3_at (endogenous control)       A     -139       A     -367       A   \n",
       "\n",
       "                                     call_31 data_32  call_32 data_33  \\\n",
       "AFFX-BioB-5_at (endogenous control)     -124       A     -135       A   \n",
       "AFFX-BioB-M_at (endogenous control)      -79       A     -186       A   \n",
       "AFFX-BioB-3_at (endogenous control)      -37       A      -70       A   \n",
       "AFFX-BioC-5_at (endogenous control)      330       A      337       A   \n",
       "AFFX-BioC-3_at (endogenous control)     -188       A     -407       A   \n",
       "\n",
       "                                                                   descr  \n",
       "AFFX-BioB-5_at (endogenous control)  AFFX-BioB-5_at (endogenous control)  \n",
       "AFFX-BioB-M_at (endogenous control)  AFFX-BioB-M_at (endogenous control)  \n",
       "AFFX-BioB-3_at (endogenous control)  AFFX-BioB-3_at (endogenous control)  \n",
       "AFFX-BioC-5_at (endogenous control)  AFFX-BioC-5_at (endogenous control)  \n",
       "AFFX-BioC-3_at (endogenous control)  AFFX-BioC-3_at (endogenous control)  \n",
       "\n",
       "[5 rows x 78 columns]"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Goal: implement differential expression analysis\n",
    "for two Leukemia conditions usig the Z-test.\n",
    "\n",
    "Notes:\n",
    "\n",
    "- The data for this analysis was downloaded from\n",
    "    http://portals.broadinstitute.org/cgi-bin/cancer/publications/pub_paper.cgi?mode=view&paper_id=43\n",
    "- This code is an adaptation for pandas of the code provided at\n",
    "    http://dept.stat.lsa.umich.edu/~kshedden/Python-Workshop/gene_expression_comparison.html\n",
    "    which uses data in \"SOFT\" format -- https://www.ncbi.nlm.nih.gov/sites/GDSbrowser?acc=GDS1615\n",
    "\n",
    "\n",
    "@author: Andrei Sura\n",
    "@see https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3134237/ - Gene Set Enrichment Analysis Made Simple\n",
    "@see http://www.gettinggeneticsdone.com/2012/03/pathway-analysis-for-high-throughput.html\n",
    "\n",
    "#########################################################\n",
    "# Expression profiling in early onset colorectal cancer\n",
    "@see http://clincancerres.aacrjournals.org/content/13/4/1107.full-text.pdf\n",
    "@see https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE4107\n",
    "\n",
    "@see http://isites.harvard.edu/fs/docs/icb.topic1517419.files/golub_analysis.R\n",
    "@see http://svitsrv25.epfl.ch/R-doc/library/multtest/doc/golub.R\n",
    "\n",
    "@see https://www.biostars.org/p/16137/\n",
    "@see http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0012336\n",
    "@see http://bioinformatics.mdanderson.org/MicroarrayCourse/Lectures09/ma10b.pdf\n",
    "@see http://sebastianraschka.com/Articles/2014_pca_step_by_step.html\n",
    "\n",
    "@see https://www.biostars.org/p/42219/ - pathway analysis\n",
    "\"\"\"\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats.distributions import norm\n",
    "# np.seterr(invalid='ignore')\n",
    "\n",
    "IN_FILE = '/Users/asura/git/gene_expression_two_group_comparison/_golub_assignment/data_set_ALL_AML_train.csv'\n",
    "OUT_FILE = '/Users/asura/git/gene_expression_two_group_comparison/_golub_assignment/out.csv'\n",
    "\n",
    "# RANGE_ALL = [\"call.{}\".format(x) for x in range(0, 27)]\n",
    "# RANGE_AML = [\"call.{}\".format(x) for x in range(27, 38)]\n",
    "RANGE_ALL = [\"call_{}\".format(x) for x in range(1, 28)]\n",
    "RANGE_AML = [\"call_{}\".format(x) for x in range(28, 39)]\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "38 samples divided in two groups:\n",
    "- ALL: 27 (1 to 27)  \n",
    "- AML: 11 (28 to 38)\n",
    "\n",
    "An accession number in bioinformatics is a unique identifier given to a DNA\n",
    "or protein sequence record to allow for tracking of different versions of\n",
    "that sequence record and the associated sequence over time in a single data\n",
    "repository. Because of its relative stability, accession numbers can be \n",
    "utilized as foreign keys for referring to a sequence object, but not necessarily\n",
    "to a unique sequence. All sequence information repositories implement the\n",
    "concept of \"accession number\" but might do so with subtle variations.\n",
    "\n",
    "\"\"\"\n",
    "df = pd.read_csv(IN_FILE, sep='\\t', skipinitialspace=False, nrows=None)\n",
    "df['descr'] = df.index\n",
    "\n",
    "# df.reset_index(inplace=True, drop=False)\n",
    "# df.reindex()\n",
    "# df.set_index(range(len(df)))\n",
    "\n",
    "# fix weird error with reading data\n",
    "del df['call_33']\n",
    "df.rename(columns={\"call\": \"call.0\", \"Gene_Accession_Number\": \"call_33\"}, inplace=True)\n",
    "# sorted(df)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>call_28</th>\n",
       "      <th>call_29</th>\n",
       "      <th>call_30</th>\n",
       "      <th>call_31</th>\n",
       "      <th>call_32</th>\n",
       "      <th>call_33</th>\n",
       "      <th>call_34</th>\n",
       "      <th>call_35</th>\n",
       "      <th>call_36</th>\n",
       "      <th>call_37</th>\n",
       "      <th>call_38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AFFX-BioB-5_at (endogenous control)</th>\n",
       "      <td>15</td>\n",
       "      <td>-318</td>\n",
       "      <td>-32</td>\n",
       "      <td>-124</td>\n",
       "      <td>-135</td>\n",
       "      <td>-214</td>\n",
       "      <td>7</td>\n",
       "      <td>-213</td>\n",
       "      <td>-25</td>\n",
       "      <td>-72</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioB-M_at (endogenous control)</th>\n",
       "      <td>-114</td>\n",
       "      <td>-192</td>\n",
       "      <td>-49</td>\n",
       "      <td>-79</td>\n",
       "      <td>-186</td>\n",
       "      <td>-153</td>\n",
       "      <td>-100</td>\n",
       "      <td>-252</td>\n",
       "      <td>-20</td>\n",
       "      <td>-139</td>\n",
       "      <td>-116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioB-3_at (endogenous control)</th>\n",
       "      <td>2</td>\n",
       "      <td>-95</td>\n",
       "      <td>49</td>\n",
       "      <td>-37</td>\n",
       "      <td>-70</td>\n",
       "      <td>-58</td>\n",
       "      <td>-57</td>\n",
       "      <td>136</td>\n",
       "      <td>124</td>\n",
       "      <td>-1</td>\n",
       "      <td>-125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioC-5_at (endogenous control)</th>\n",
       "      <td>193</td>\n",
       "      <td>312</td>\n",
       "      <td>230</td>\n",
       "      <td>330</td>\n",
       "      <td>337</td>\n",
       "      <td>88</td>\n",
       "      <td>132</td>\n",
       "      <td>318</td>\n",
       "      <td>325</td>\n",
       "      <td>392</td>\n",
       "      <td>241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BioC-3_at (endogenous control)</th>\n",
       "      <td>-51</td>\n",
       "      <td>-139</td>\n",
       "      <td>-367</td>\n",
       "      <td>-188</td>\n",
       "      <td>-407</td>\n",
       "      <td>-295</td>\n",
       "      <td>-377</td>\n",
       "      <td>-209</td>\n",
       "      <td>-396</td>\n",
       "      <td>-324</td>\n",
       "      <td>-191</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     call_28  call_29  call_30  call_31  \\\n",
       "AFFX-BioB-5_at (endogenous control)       15     -318      -32     -124   \n",
       "AFFX-BioB-M_at (endogenous control)     -114     -192      -49      -79   \n",
       "AFFX-BioB-3_at (endogenous control)        2      -95       49      -37   \n",
       "AFFX-BioC-5_at (endogenous control)      193      312      230      330   \n",
       "AFFX-BioC-3_at (endogenous control)      -51     -139     -367     -188   \n",
       "\n",
       "                                     call_32  call_33  call_34  call_35  \\\n",
       "AFFX-BioB-5_at (endogenous control)     -135     -214        7     -213   \n",
       "AFFX-BioB-M_at (endogenous control)     -186     -153     -100     -252   \n",
       "AFFX-BioB-3_at (endogenous control)      -70      -58      -57      136   \n",
       "AFFX-BioC-5_at (endogenous control)      337       88      132      318   \n",
       "AFFX-BioC-3_at (endogenous control)     -407     -295     -377     -209   \n",
       "\n",
       "                                     call_36  call_37  call_38  \n",
       "AFFX-BioB-5_at (endogenous control)      -25      -72       -4  \n",
       "AFFX-BioB-M_at (endogenous control)      -20     -139     -116  \n",
       "AFFX-BioB-3_at (endogenous control)      124       -1     -125  \n",
       "AFFX-BioC-5_at (endogenous control)      325      392      241  \n",
       "AFFX-BioC-3_at (endogenous control)     -396     -324     -191  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = [\"descr\"]\n",
    "columns_data = []\n",
    "columns_data.extend(RANGE_ALL)\n",
    "columns_data.extend(RANGE_AML)\n",
    "\n",
    "columns.extend(RANGE_ALL)\n",
    "columns.extend(RANGE_AML)\n",
    "df = df[columns]\n",
    "# df.describe()\n",
    "# sorted(df)\n",
    "# df.head(10)\n",
    "df_all = df[RANGE_ALL]\n",
    "df_aml = df[RANGE_AML]\n",
    "df_aml.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Z mean: 0.140047655782809\n",
      "Z std: 1.6058829875390022\n",
      "Z means: 0.22008696871931546, 0.13073362322906437, 0.8692663767709357\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# df_all = df_all.applymap(lambda x: np.log(x) / np.log(2))\n",
    "# df_aml = df_aml.applymap(lambda x: np.log(x) / np.log(2))\n",
    "\n",
    "mean_all = df_all.mean(axis=1)  # means for 7129 genes\n",
    "mean_aml = df_aml.mean(axis=1)   \n",
    "var_all = df_all.var(axis=1)  # variance for 7129 genes\n",
    "var_aml = df_aml.var(axis=1)\n",
    "\n",
    "num_all = len(RANGE_ALL)  # number of samples of each condition\n",
    "num_aml = len(RANGE_AML)\n",
    "\n",
    "Z = (mean_all - mean_aml) / np.sqrt(var_all/num_all + var_aml/num_aml)\n",
    "# print(\"means ALL: {}\\n {}\".format(len(mean_all), mean_all))\n",
    "# print(\"means AML: {}\\n {}\".format(len(mean_aml), mean_aml))\n",
    "\n",
    "print(\"Z mean: {}\".format(Z.mean()))\n",
    "print(\"Z std: {}\".format(Z.std()))\n",
    "\n",
    "\"\"\"\n",
    "Z mean: 0.14004765578280895\n",
    "Z std: 1.6058829875390022\n",
    "Z means: 0.22008696871931546, 0.13073362322906437, 0.8692663767709357\n",
    "\n",
    "Since the standard deviation is greater than 1, there appear to be multiple\n",
    "genes for which the mean expression levels in the ALL and AML samples differ. \n",
    "Further, since the mean Z-score is positive, it appears that the dominant pattern\n",
    "is for genes to be expressed at a higher level in the ALL compared to the AML samples.\n",
    "\"\"\"\n",
    "\n",
    "mean_z_mod = np.mean(np.abs(Z) > 2)\n",
    "mean_z1 = np.mean(Z > 2)\n",
    "mean_z2 = np.mean(Z < 2)\n",
    "\n",
    "print(\"Z means: {}, {}, {}\".format(mean_z_mod, mean_z1, mean_z2))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Check the mean of values with low Z magnitude: 0.21052631578947367\n",
      "Check the mean of values with high Z magnitude: 0.21052631578947367\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Genes with low expression level are harder to measure accurately, \n",
    "thus we expect that fewer of these genes will meet a given statistical\n",
    "threshold for differential expression. Similarly, genes with low variance\n",
    "are potentially less likely to be associated with biological differences.\n",
    "We can assess these trends in our data. First, we will determine which\n",
    "genes are in the lower and upper halves of all genes based on either \n",
    "mean expression or expression variation.\n",
    "\"\"\"\n",
    "SD = df[columns_data].std()\n",
    "index_stdev_low = np.flatnonzero(SD < np.median(SD))\n",
    "index_stdev_high = np.flatnonzero(SD > np.median(SD))\n",
    "\n",
    "MEAN = df[columns_data].mean()\n",
    "index_mean_low = np.flatnonzero(MEAN < np.median(MEAN))\n",
    "index_mean_high = np.flatnonzero(MEAN > np.median(MEAN))\n",
    "\n",
    "\"\"\"\n",
    "Now we can look at the proportion of genes within each of these\n",
    "strata that have Z-score magnitude greater than two.\n",
    "\"\"\"\n",
    "mean_low = np.mean(np.abs(Z[index_stdev_low]) > 2)\n",
    "mean_high = np.mean(np.abs(Z[index_stdev_high]) > 2)\n",
    "\n",
    "print(\"Check the mean of values with low Z magnitude: {}\".format(mean_low))\n",
    "print(\"Check the mean of values with high Z magnitude: {}\".format(mean_high))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "148 - Clone 22 mRNA, alternative splice variant alpha-1\n",
      "531 - HMG1 High-mobility group (nonhistone chromosomal) protein 1\n",
      "1134 - ASPARTYL-TRNA SYNTHETASE\n",
      "1143 - SPTAN1 Spectrin, alpha, non-erythrocytic 1 (alpha-fodrin)\n",
      "1206 - Protein tyrosine kinase related mRNA sequence\n",
      "1244 - GBE1 Glucan (1,4-alpha-), branching enzyme 1 (glycogen branching enzyme, Andersen disease, glycogen storage disease type IV)\n",
      "1629 - Inducible protein mRNA\n",
      "1703 - ADA Adenosine deaminase\n",
      "1810 - MDU1 Antigen identified by monoclonal antibodies 4F2, TRA1.10, TROP4, and T43\n",
      "1923 - PRKAR2B Protein kinase, cAMP-dependent, regulatory, type II, beta\n",
      "2019 - FAH Fumarylacetoacetate\n",
      "2300 - ACTN2 Actinin alpha 2 \n",
      "2440 - HKR-T1\n",
      "2625 - MSH2 DNA repair protein MSH2\n",
      "2758 - Thrombospondin-p50 gene extracted from Human thrombospondin-1 gene, partial cds\n",
      "3069 - Tax1-binding protein TXBP181 mRNA\n",
      "3311 - Protein kinase ATR mRNA\n",
      "3319 - Leukotriene C4 synthase (LTC4S) gene\n",
      "3846 - GB DEF = Homeodomain protein HoxA9 mRNA\n",
      "4229 - CTPS CTP synthetase\n",
      "4327 - PROTEASOME IOTA CHAIN\n",
      "4374 - DAGK1 Diacylglycerol kinase, alpha (80kD)\n",
      "4534 - RETINOBLASTOMA BINDING PROTEIN P48\n",
      "5071 - Cytokeratin 17\n",
      "5190 - Adenosine triphosphatase, calcium\n",
      "5771 - C-myb gene extracted from Human (c-myb) gene, complete primary cds, and five complete alternatively spliced cds\n",
      "6293 - GB DEF = Type A plasminogen related gene, exon 3 and complete cds\n",
      "6514 - DHPS Deoxyhypusine synthase\n",
      "6643 - Butyrophilin (BTF5) mRNA\n"
     ]
    }
   ],
   "source": [
    "## The Z-score threshold under a Bonferroni correction\n",
    "zst = -norm.ppf(0.025/Z.shape[0])\n",
    "indexes = np.flatnonzero(np.abs(Z) > zst)\n",
    "\n",
    "with open(OUT_FILE, \"w\") as fout:\n",
    "    for i in indexes:\n",
    "        # print(\"Found gene: {} - {}\".format(i, df[\"Gene Description\"][i]))\n",
    "        print(\"{} - {}\".format(i, df[\"descr\"][i]))\n",
    "        fout.write(\"{},{}\\n\".format(Z[i], df[\"descr\"][i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
